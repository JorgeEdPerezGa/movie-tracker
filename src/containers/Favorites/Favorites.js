import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import MovieCard from '../../components/MovieCard/MovieCard';
import { addFavorite, removeFavorite, updateMovies } from '../../actions/';
import { postFavorite, deleteFavorite } from '../../helper';
import './Favorites.css';

export class Favorites extends Component {

  renderFavorites = () => {
    return this.props.favorites.map((movie, idx) => {
      return <MovieCard key={movie.movie_id + idx} movie={movie} onFavorite={this.handleFavorites}/>;
    });
  }

   handleFavorites = (movie) => {
    const duplicated = this.props.favorites.some(fav => movie.title === fav.title);

    const favMovie = {...movie, favorite: !movie.favorite};

    duplicated ? this.removeFavMovie(favMovie) :this.addFavMovie(favMovie);

    this.props.updateMovies(favMovie);
  }

  addFavMovie = (movie) => {
    this.props.addFavorite(movie);
    postFavorite(movie, this.props.user);
  }

  removeFavMovie = (movie) => {
    this.props.removeFavorite(movie);
    deleteFavorite(movie, this.props.user)
  }

  render() {
    return (
      <div className='movie-list'>
        { this.renderFavorites() }
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  favorites: state.favorites,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  addFavorite: movie => dispatch(addFavorite(movie)),
  removeFavorite: movie => dispatch(removeFavorite(movie)),
  updateMovies: movie => dispatch(updateMovies(movie))
})

Favorites.propTypes = {
  favorites: PropTypes.array,
  user: PropTypes.object,
  addFavorite: PropTypes.func,
  removeFavorite: PropTypes.func,
  updateMovies: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
