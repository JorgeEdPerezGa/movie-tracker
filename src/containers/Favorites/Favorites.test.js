import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { Favorites } from './Favorites';

describe('Favorites', () => {
  it.skip('should not pass', () => {
    expect(false).toEqual(true);
  })

  let mockFavorites;
  let mockUser;
  let mockAddFavorite;
  let mockRemoveFavorite;
  let mockUpdateMovies;
  let renderedComponent;
  
  beforeAll(() => {
    mockFavorites = [{
      backdrop_path: "/askg3SMvhqEl4OL52YuvdtY40Yb.jpg",
      movie_id: 354912,
      overview: "Despite his familyâ€™s baffling generations-old ban on music",
      poster_path: "/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg",
      release_date: "2017-10-27",
      title: "Coco",
      vote_average: 7.7,
      favorite: true
    }];
    
    mockUser = {
      email: "fake@123.com",
      id: 2,
      name: "ted",
      password: "pw123"
    };
  })

  beforeEach(() => {
    mockRemoveFavorite = jest.fn();
    mockUpdateMovies = jest.fn();
    renderedComponent = jest.fn();
    renderedComponent = shallow(
      <Favorites 
        favorites={mockFavorites}
        user={mockUser}
        addFavorite={mockAddFavorite}
        removeFavorite={mockRemoveFavorite}
        updateMovies={mockUpdateMovies}
      />
    );
  })

  it.skip('should match snapshot', () => {
    expect(renderedComponent).toMatchSnapshot();
  })

  it('when user favorites a card and it is already in Favorites, it is removed', () => {

  })

  it('when user favorites a card and it is not in Favorites, it is added', () => {

  })
})