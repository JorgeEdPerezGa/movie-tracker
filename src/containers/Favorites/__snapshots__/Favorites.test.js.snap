// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    addFavorite={[Function]}
    favorites={
        Array [
            Object {
              "backdrop_path": "/def.jpg",
              "favorite": true,
              "movie_id": 999,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/abs.jpg",
              "release_date": "2017-10-27",
              "title": "Thor",
              "vote_average": 7.2,
            },
          ]
    }
    removeFavorite={[Function]}
    updateMovies={[Function]}
    user={
        Object {
            "email": "fake@123.com",
            "id": 2,
            "name": "ted",
            "password": "pw123",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MovieCard
          movie={
                    Object {
                              "backdrop_path": "/def.jpg",
                              "favorite": true,
                              "movie_id": 999,
                              "overview": "Despite his family’s baffling generations-old ban on music",
                              "poster_path": "/abs.jpg",
                              "release_date": "2017-10-27",
                              "title": "Thor",
                              "vote_average": 7.2,
                            }
          }
          onFavorite={[Function]}
/>,
      ],
      "className": "movie-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "999",
        "nodeType": "function",
        "props": Object {
          "movie": Object {
            "backdrop_path": "/def.jpg",
            "favorite": true,
            "movie_id": 999,
            "overview": "Despite his family’s baffling generations-old ban on music",
            "poster_path": "/abs.jpg",
            "release_date": "2017-10-27",
            "title": "Thor",
            "vote_average": 7.2,
          },
          "onFavorite": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MovieCard
            movie={
                        Object {
                                    "backdrop_path": "/def.jpg",
                                    "favorite": true,
                                    "movie_id": 999,
                                    "overview": "Despite his family’s baffling generations-old ban on music",
                                    "poster_path": "/abs.jpg",
                                    "release_date": "2017-10-27",
                                    "title": "Thor",
                                    "vote_average": 7.2,
                                  }
            }
            onFavorite={[Function]}
/>,
        ],
        "className": "movie-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "999",
          "nodeType": "function",
          "props": Object {
            "movie": Object {
              "backdrop_path": "/def.jpg",
              "favorite": true,
              "movie_id": 999,
              "overview": "Despite his family’s baffling generations-old ban on music",
              "poster_path": "/abs.jpg",
              "release_date": "2017-10-27",
              "title": "Thor",
              "vote_average": 7.2,
            },
            "onFavorite": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
